[BLOGPOST] Instant DeepSeek: One-Click Activation with OpenSearch (#3602)
opensearch-project/project-website: _posts/2025-01-30-one-click-deepseek-integration.md

### Without Flow Framework plugin

In the ([previous blog](https://opensearch.org/blog/OpenSearch-Now-Supports-DeepSeek-Chat-Models/)) blog post, setting up DeepSeek model or any LLM model involves four separate API calls:

1. Create a connector for DeepSeek model
2. Create a model group
3. Register the model using the connector ID
4. Create a search pipeline for RAG

Via OpenSearch Flow Framework plugin, we are simplifying this experience by enabling you to make a single API call. Let’s demonstrate how the plugin simplifies this process using the preceding conversational search example.

### With the Flow Framework plugin

> In the following example, you will configure the  conversational_search_with_llm_deploy  workflow template to implement RAG with DeepSeek in OpenSearch. The workflow created using this template performs following configuration steps:

Reviewer: remove 1 spacing both before and after "conversational_search_with_llm_deploy". Also, bold the template name

Reviewer: should be as:  In the following example, you will configure the **conversational_search_with_llm_deploy** workflow template to implement RAG with DeepSeek in OpenSearch. The workflow created using this template performs following configuration steps:

* Deploys an externally hosted DeepSeek model
* Registers and deploys the model
* Creates a search pipeline with RAG processor

### Step 1: Create and provision the workflow

> Using the conversational_search_with_llm_deploy  workflow template, you can provision the workflow with required fields—the API key for the DeepSeek model:

Reviewer: remove extra spacing between "conversational_search_with_llm_deploy"  and "workflow"

---
layout: post
title:  "Instant DeepSeek: One-click activation with OpenSearch"
authors:
   - kazabdu
   - minalsha
   - seanzheng
   - amitgalitz
   - kolchfa
date: 2025-01-31
categories:
  - technical-posts
meta_keywords: OpenSearch DeepSeek integration, LLM integration, RAG, AI search, Flow Framework, machine learning, natural language processing, open-source LLM
meta_description: Learn how to enable DeepSeek LLM integration in OpenSearch with just one API call using the Flow Framework plugin, simplifying the setup process for conversational search and RAG.
---

> In an [earlier blog post](https://opensearch.org/blog/OpenSearch-Now-Supports-DeepSeek-Chat-Models/), we introduced OpenSearch's support for the DeepSeek LLM. This post focuses on simplifying DeepSeek LLM integration using the OpenSearch Flow Framework plugin. With just one API call, you can provision the entire integration—creating connectors, registering models, deploying them, and setting up agents and tools. Automated templates handle the setup, eliminating the need to call multiple APIs or manage complex orchestration.

Reviewer 2: ```suggestion
In an [earlier blog post](https://opensearch.org/blog/OpenSearch-Now-Supports-DeepSeek-Chat-Models/), we introduced OpenSearch's support for the DeepSeek large language model (LLM). This post focuses on simplifying DeepSeek LLM integration using the OpenSearch Flow Framework plugin. With just one API call, you can provision the entire integration—creating connectors, registering models, deploying them, and setting up agents and tools. Automated templates handle the setup, eliminating the need to call multiple APIs or manage complex orchestration.
```

## Manual setup

> In our [earlier blog post](https://opensearch.org/blog/OpenSearch-Now-Supports-DeepSeek-Chat-Models/), setting up the DeepSeek model—--or any LLM—--required four separate API calls:  

Reviewer 2: ```suggestion
In our [earlier blog post](https://opensearch.org/blog/OpenSearch-Now-Supports-DeepSeek-Chat-Models/), setting up the DeepSeek model—or any LLM—required four separate API calls:  
```

1. Creating a connector for the DeepSeek model  
2. Creating a model group  
3. Registering the model using the connector ID  

> 4. Creating a search pipeline for RAG  

Reviewer 2: ```suggestion
4. Creating a search pipeline for retrieval-augmented generation (RAG)  
```

With the OpenSearch Flow Framework plugin, this process is now streamlined into a single API call. In the following example, we'll present a simplified setup using the conversational search example from the earlier blog post.

## One-click deployment

> In the following example, you will configure the `conversational_search_with_llm_deploy` workflow template to implement RAG with DeepSeek in OpenSearch. The workflow created using this template performs following configuration steps:

Reviewer 2: ```suggestion
In the following example, you will configure the `conversational_search_with_llm_deploy` workflow template to implement RAG with DeepSeek in OpenSearch. The workflow created using this template performs the following configuration steps:
```

Create a conversation memory to store all messages from a conversation:

```json
POST /_plugins/_ml/memory/
{
"name": "Conversation about NYC population"
}
```

The response contains a memory ID for the created memory:

```json
{
"memory_id": "znCqcI0BfUsSoeNTntd7"
}
```

### Step 3: Use the pipeline for RAG

Send a query to OpenSearch and provide additional parameters in the `ext.generative_qa_parameters` object:

```json
GET /my_rag_test_data/_search
{
  "query": {
    "match": {
      "text": "What's the population of NYC metro area in 2023"
    }
  },
  "ext": {
    "generative_qa_parameters": {
      "llm_model": "deepseek-chat",
      "llm_question": "What's the population of NYC metro area in 2023",
      "memory_id": "znCqcI0BfUsSoeNTntd7", <can skip memory_id if skipped step2>
      "context_size": 5,
      "message_size": 5,
      "timeout": 15
    }
  }
}
```

The response contains the model output:

```json
{
  ...
  "ext": {
    "retrieval_augmented_generation": {
      "answer": "The population of the New York City metro area in 2023 was 18,867,000.",
      "message_id": "p3CvcI0BfUsSoeNTj9iH"
    }
  }
}
```

## Additional use cases

> The example we've covered is just one of many possible workflows. Flow Framework comes with a variety of pre-built templates designed for different scenarios. You can explore our [substitution templates](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/substitutionTemplates) for various workflows and review their corresponding [default configurations](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/defaults).

Reviewer 2: ```suggestion
The preceding example represents just one of many possible workflows. The Flow Framework plugin comes with a variety of prebuilt templates designed for different scenarios. You can explore our [substitution templates](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/substitutionTemplates) for various workflows and review their corresponding [default configurations](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/defaults).
```

These resources will help you discover and implement other automated workflows that best suit your needs.

## Conclusion

> By using the Flow Framework plugin, we've transformed a complex, multi-step setup process into a simple, single API call. This simplification isn't limited to DeepSeek—--you can use the same streamlined approach to deploy models from other leading LLM providers like Cohere and OpenAI. Whether you're experimenting with different models or setting up production environments, the Flow Framework makes LLM integration faster and more reliable.

Reviewer 2: ```suggestion
By using the Flow Framework plugin, we've transformed a complex, multi-step setup process into a single, simple API call. This simplification isn't limited to DeepSeek—you can use the same streamlined approach to deploy models from other leading LLM providers like Cohere and OpenAI. Whether you're experimenting with different models or setting up production environments, the Flow Framework plugin makes LLM integration faster and more reliable.
```
